app/drivers/kscan/debounce.c

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright (c) 2021 The ZMK Contributors
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#include "debounce.h"
+
+static uint32_t get_debounce_cycles(const struct debounce_state *state,
+                                    const struct debounce_config *config) {
+    return state->pressed ? config->debounce_up_ms : config->debounce_down_ms;
+}
+
+bool debounce_update(struct debounce_state *state, const bool active,
+                     const struct debounce_config *config) {
+    // This uses a variation of the integrator debouncing described at
+    // https://www.kennethkuhn.com/electronics/debounce.c
+    // We may have different thresholds for press and release, so we reset the
+    // counter and begin counting up again every time the state flips.
```

Comment:
Since even with the comment in the link as reference it took me a little bit to figure out how this works, might be worth adding a little more description to the comment here. This is how I've been thinking about it if it's helpful:

> There is a counter that represents how close we are to transitioning away from the current state--it starts at 0. On every tick, if the new state (`active`) is different from the current state we increment the counter, otherwise we decrement it. When the counter reaches the configured threshold (can be different for key down vs. up) we transition to the new state and reset the counter back to zero.